File: service/webserver/src/main/java/com/hagoapp/datacova/Application.java
/*
 * Copyright (c) 2020.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 */

package com.hagoapp.datacova;

import com.hagoapp.datacova.command.Start;
import com.hagoapp.datacova.command.Stop;
import picocli.CommandLine;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * The entry of application. Based on command line arguments, user can start / stop services or execute a task.
 */
@CommandLine.Command(name = "DataCoVa", version = "0.1", description = "Service of DataCoVa", subcommands = {
        Start.class, Stop.class
})
public class Application {

    private static final Map<String, Object> internalData = new ConcurrentHashMap<>();

    public static void main(String[] args) {
        var app = new Application();
        var cli = new CommandLine(app);
        cli.setExecutionStrategy(app::executionStrategy).execute(args);
    }

    public static void setData(String key, Object data) {
        internalData.put(key, data);
    }

    public static Object getData(String key) {
        return internalData.get(key);
    }

    private int executionStrategy(CommandLine.ParseResult parseResult) {
        return new CommandLine.RunLast().execute(parseResult); // default execution strategy
    }
}

File: service/dispatcher/src/main/kotlin/com/hagoapp/datacova/dispatcher/Application.kt
/*
 * Copyright (c) 2021.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.hagoapp.datacova.dispatcher

import com.google.gson.Gson
import com.hagoapp.datacova.lib.data.TaskExecutionData
import com.hagoapp.datacova.lib.execution.TaskExecution
import org.slf4j.LoggerFactory
import java.io.File
import java.io.FileInputStream
import java.nio.charset.StandardCharsets
import java.util.concurrent.atomic.AtomicBoolean

class Application {
    companion object {

        lateinit var config: Config
            private set
            @JvmStatic
            get
        private var exitFlag = AtomicBoolean(false)
        private val logger = LoggerFactory.getLogger(Application::class.java)
        private const val DEFAULT_CONFIG_FILE = "conf.json"

        @JvmStatic
        fun main(args: Array<String>) {
            val configFile = if (args.isNotEmpty()) args[0] else DEFAULT_CONFIG_FILE
            if (!File(configFile).exists()) {
                logger.error("config file not found from {}, exit", configFile)
                return
            }
            config = FileInputStream(configFile).use {
                val json = String(it.readAllBytes(), StandardCharsets.UTF_8)
                Gson().fromJson(json, Config::class.java)
            }
            Thread { DispatchServer.start() }.start()
            while (!exitFlag.get()) {
                /*val l = loadActiveTaskExecution()
                if (l.isEmpty()) {
                    Thread.sleep(5000)
                    continue
                }*/
            }
            logger.info("exit")
        }

        private fun loadActiveTaskExecution(): List<TaskExecution> {
            return TaskExecutionData(config.db).use {
                it.loadQueueingTaskExecution()
            }
        }
    }
}
